### ISSUE TITLE
persistent storage isn't available due to error "The specified item already exists in the keychain."

### ISSUE THREAD

**Issue by mateuszlitwin** [Azure/azure-sdk-for-go#23588]
### Bug Report

- output of `go version`

```
	github.com/Azure/azure-sdk-for-go/sdk/azidentity v1.8.0
	github.com/Azure/azure-sdk-for-go/sdk/azidentity/cache v0.3.0
```

- What happened?

Executing *multiple* instances of a program on macOS with keychain token caching (`"github.com/Azure/azure-sdk-for-go/sdk/azidentity/cache"`) fails with:

```
persistent storage isn't available due to error "The specified item already exists in the keychain. (-25299)"
```

This is because the `tryStorage` function is executed many times when many instances of a program are executed concurrently: https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/cache/cache.go#L71 The `sync.Once` is obviously not enough to prevent cache from "trying" storage simultaneously from many programs at once.

- What did you expect or want to happen?

Cache creation not failing.

- How can we reproduce it?

Execute program with token caching many times in parallel. I think just many `cache.New(nil)` in parallel is needed.

```
package main

import (
	"fmt"

	"github.com/Azure/azure-sdk-for-go/sdk/azidentity/cache"
)

func main() {
	_, err := cache.New(nil)
	if err != nil {
		fmt.Printf("failed to create cache: %s", err)
	}
}

```

```
for i in {1..10}; do ./main &; done; wait
```

- Anything we should know about your environment.

I suggest giving an option to disable the `tryStorage` check in `cache.Options`. I think the change is trivial and seems to work in my environment.

Additionally, the `tryStorage` takes ~30ms on M2 MAX so it would be nice to have an option to shave that time from the Go cli.

---

General question: Is the cache safe for concurrent use from multiple programs? e.g. requesting and caching tokens from multiple programs simultaneously on macOS? can keychain be modified concurrently?

**Comment by chlowell** [comment]
Thanks for opening this issue! You're quite right that `cache.New` does nothing to prevent conflict between multiple processes, and it should. I think we don't want to synchronize processes with e.g. a kernel mutex but we can prevent contention by simply adding some randomness to the test cache name. That should unblock all the common scenarios, although on at least some platforms it would still be possible for processes to fail the storage test if the degree of concurrency were very high. How many simultaneous processes do you need?

As for making the storage test optional, I agree that's trivial and would also solve this problem. However, I think it would make the API more difficult to understand and the broader feature harder to use:
```go
// before
c, err := cache.New(nil)
if err != nil {
    // Persistence isn't possible. Fall back to in-memory caching or
    // abandon the operation. Falling back is simple because you
    // haven't created a credential or client yet.
}

// after
c, err := cache.New(&cache.Options{
        // how do I decide whether to set this?
        DisableStorageTest: true,
    }
)
if err != nil {
    // actually err is always nil, and that doesn't indicate the cache is useable
}
// later, somewhere else in the app...
err := client.Foo()
if err != nil {
    // If this is a persistence error and you want to fall back to in-memory
    // caching, you must create a new credential and client, then call Foo again
}
```

**Comment by mateuszlitwin** [comment]
Adding randomness to the cache test is fine, but I am still worried about race conditions in other scenarios. E.g. when tokens are stored (on refresh) are and retrieved. Is cache test is unsafe, doesn't it mean that general usage is unsafe?
